OptimoveSDK/Sources/Classes/Optimobile/Optimobile+Push.swift

```diff
@@ -404,55 +404,61 @@ class PushHelper {
         let kumulosDidFailToRegister = imp_implementationWithBlock(didFailToRegBlock as Any)
         existingDidFailToReg = class_replaceMethod(klass, didFailToRegisterSelector, kumulosDidFailToRegister, didFailToRegType)
 
-        // iOS9 did receive remote delegate
         // iOS9+ content-available handler
         let didReceiveSelector = #selector(UIApplicationDelegate.application(_:didReceiveRemoteNotification:fetchCompletionHandler:))
         let receiveType = NSString(string: "v@:@@@?").utf8String
         let didReceive : didReceiveBlock = { (obj:Any, _ application: UIApplication, userInfo: [AnyHashable : Any], completionHandler: @escaping (UIBackgroundFetchResult) -> Void) in
-            var fetchResult : UIBackgroundFetchResult = .noData
-            let fetchBarrier = DispatchSemaphore(value: 0)
-
-            if let _ = existingDidReceive {
-                unsafeBitCast(existingDidReceive, to: kumulos_applicationDidReceiveRemoteNotificationFetchCompletionHandler.self)(obj, didReceiveSelector, application, userInfo, { (result : UIBackgroundFetchResult) in
-                    fetchResult = result
-                    fetchBarrier.signal()
-                })
-            } else {
-                fetchBarrier.signal()
-            }
-            
-            if UIApplication.shared.applicationState == .inactive {
-               if #available(iOS 10, *) {
-                   // Noop (tap handler in delegate will deal with opening the URL)
-               } else {
-                   Optimobile.sharedInstance.pushHandleOpen(withUserInfo:userInfo)
-               }
+            let notification = PushNotification(userInfo: userInfo)
+            let hasInApp = notification.inAppDeepLink() != nil
+
+            self.setBadge(userInfo: userInfo)
+            self.trackPushDelivery(notification: notification)
+
+            if existingDidReceive == nil && !hasInApp {
+                // Nothing to do
+                completionHandler(.noData)
+                return
+            } else if existingDidReceive != nil && !hasInApp {
```

Comment:
We could chop this escape hatch but seemed less overhead to have it at the cost of more verbosity.